src/dataflow/src/source/kafka.rs

```diff
@@ -390,7 +403,24 @@ impl KafkaSourceReader {
                     old: self.last_stats.take(),
                     new: Some(stats.clone()),
                 });
-                self.last_stats = Some(stats);
+                self.last_stats = Some(stats.clone());
+            }
+
+            // This temp var is required to satisfy the type checker
+            let res: Result<Statistics, serde_json::Error> = serde_json::from_str(&stats.to_string());
+            match res {
+                Ok(statistics) => {
+                    statistics.topics.get(&self.topic_name).and_then(|topic| {
+                        topic.partitions.iter().for_each(|(id, partition)| {
+                            self.partition_metrics
+                                .set_offset_max(*id, partition.hi_offset);
+                        });
+                        Some(())
```

Comment:
Using combinators for side effects like this is relatively non-idiomatic. I think you just want

```rust
if let Some(topic) = statistics.topics.get(&self.topic_name) {

}
```

Similarly for the `for_each` method; it's considered more idiomatic to use a loop:

> This is equivalent to using a [for](https://doc.rust-lang.org/stable/book/ch03-05-control-flow.html#looping-through-a-collection-with-for) loop on the iterator, although break and continue are not possible from a closure. It’s generally more idiomatic to use a for loop, but for_each may be more legible when processing items at the end of longer iterator chains.